Support Ecto.Query.API.json_extract_path/2

wojtekmach opened the pull request:
Ref https://github.com/elixir-ecto/ecto/pull/3227

michalmuskala commented on lib/ecto/adapters/myxql/connection.ex:
```diff
@@ -551,6 +551,19 @@ if Code.ensure_loaded?(MyXQL) do
+      path =
+        Enum.map(path, fn
+          binary when is_binary(binary) ->
+            [".", binary]
```
Do we need to do some sort of escaping here?

wojtekmach replied:
Yes, see: https://github.com/elixir-ecto/ecto/pull/3227#discussion_r377126347

michalmuskala commented on lib/ecto/adapters/postgres/connection.ex:
```diff
@@ -569,6 +569,15 @@ if Code.ensure_loaded?(Postgrex) do
+    defp expr({:json_extract_path, _, [expr, path]}, sources, query) do
+      path =
+        path
+        |> Enum.map(&to_string/1)
```
Same in here with escaping

wojtekmach commented:
@michalmuskala You were right, the way to go was properly escaping strings that ended up as JSON keys and not trying to parameterise them. Both adapters had an `escape_string`, I could have adjusted that but decided to introduce a similar `escape_json_key` function that does the same + more. I've updated this (and the other PR) accordingly and added some tests around it in Ecto integration tests. Feedback welcome!

josevalim commented:
We need some unit tests and then this is ready to go!

wojtekmach commented:
There was a bug that I realised when writing the unit tests, thank you!
